kotlinx-coroutines-core/common/src/flow/Builders.kt

```diff
@@ -315,3 +318,36 @@ private class ChannelFlowBuilder<T>(
     override fun toString(): String =
         "block[$block] -> ${super.toString()}"
 }
+
+private class CallbackFlowBuilder<T>(
+    private val block: suspend ProducerScope<T>.() -> Unit,
+    context: CoroutineContext = EmptyCoroutineContext,
+    capacity: Int = BUFFERED
+) : ChannelFlowBuilder<T>(block, context, capacity) {
+
+    private val collectCallback: suspend (ProducerScope<T>) -> Unit = {
+        collectTo(it)
+        /*
+         * We expect user either call `awaitClose` from within a block (then the channel is closed at this moment)
+         * or being closed/cancelled externally/manually. Otherwise "user forgot to call
+         * awaitClose and receives unhelpful ClosedSendChannelException exceptions" situation is detected.
+         */
+        if (it.isActive && !it.isClosedForSend) {
+            throw IllegalStateException(
+                """
+                    'awaitClose { yourCallbackOrListener.cancel() }' should be used in the end of callbackFlow block.
+                    Otherwise, a callback/listener may leak in case of cancellation external cancellation (e.g. by 'take(1)' or destroyed activity).
+                    For a more detailed explanation, please refer to callbackFlow KDoc. 
+            """.trimIndent())
```

Comment:
Style nit: one more tab here. 

Also "KDoc" does not read nicely in prose. I'd replace it "... callbackFlow API documentation." And this can be much shorter sentence giving the same idea: "See callbackFlow API documentation for details".